File: lint-arwaky/archive/infrastructure/javascript_naming_provider.py
"""JSTracer - Naming variants for JavaScript/TypeScript symbols."""

from __future__ import annotations

import re
from typing import cast
from ..taxonomy import NameVariants, SymbolName
from ..contract import INamingProviderPort


class JavascriptNamingProvider(INamingProviderPort):
    """Implementation of naming variants for JavaScript/TypeScript symbols."""

    def get_variants(self, name: SymbolName) -> NameVariants:
        """Generate common naming variants (snake, camel, pascal, screaming, kebab)."""
        name_str = str(name)
        d = self._get_variant_dict(name_str)
        kebab = d["snake_case"].replace("_", "-")
        variants = {
            name_str,
            d["snake_case"],
            d["camel_case"],
            d["pascal_case"],
            d["screaming_snake"],
            kebab,
        }
        return NameVariants(values=list(variants))

    def _get_variant_dict(self, name: str) -> dict:
        """Produce common naming variants for a symbol name."""
        words = re.findall(r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+", name)
        words = cast(list[str], [w.lower() for w in words])
        if not words:
            return {
                "snake_case": name,
                "camel_case": name,
                "pascal_case": name,
                "screaming_snake": name,
            }
        snake_case = "_".join(words)
        first = str(words[0])
        rest = "".join(str(w).capitalize() for w in words[1:])
        return {
            "snake_case": snake_case,
            "camel_case": first + rest,
            "pascal_case": "".join(str(w).capitalize() for w in words),
            "screaming_snake": snake_case.upper(),
        }

File: lint-arwaky/src-rust/infrastructure/javascript_naming_provider.rs
/// javascript_naming_provider — Naming variants for JavaScript/TypeScript symbols.
use crate::contract::INamingProviderPort;
use crate::taxonomy::{NameVariants, SymbolName};
use regex::Regex;

pub struct JavascriptNamingProvider;

impl JavascriptNamingProvider {
    pub fn new() -> Self {
        Self
    }
}

#[async_trait::async_trait]
impl INamingProviderPort for JavascriptNamingProvider {
    async fn get_variants(&self, name: &SymbolName) -> Result<NameVariants, crate::taxonomy::NamingError> {
        let name_str = &name.value;
        let words: Vec<String> = Regex::new(r"[A-Za-z][a-z0-9]*|[A-Z]+(?=[A-Z][a-z0-9]|\b)|[0-9]+")
            .unwrap().find_iter(name_str).map(|m| m.as_str().to_lowercase()).collect();
        if words.is_empty() {
            return Ok(NameVariants::new(vec![name_str.clone()]));
        }
        let snake_case = words.join("_");
        let first = words[0].clone();
        let rest: String = words[1..].iter().map(|w| {
            let mut c = w.chars();
            match c.next() { Some(ch) => ch.to_uppercase().to_string() + c.as_str(), None => String::new() }
        }).collect();
        let camel_case = format!("{}{}", first, rest);
        let pascal_case: String = words.iter().map(|w| {
            let mut c = w.chars();
            match c.next() { Some(ch) => ch.to_uppercase().to_string() + c.as_str(), None => String::new() }
        }).collect();
        let screaming_snake = snake_case.to_uppercase();
        let kebab = snake_case.replace('_', "-");
        let mut variants = vec![name_str.clone(), snake_case, camel_case, pascal_case, screaming_snake, kebab];
        variants.sort();
        variants.dedup();
        Ok(NameVariants::new(variants))
    }
}
